File: packages/place-core/clients/kotlin/GoldenClient.kt
/*
 * Synthetic UniFFI test client (Kotlin / JVM).
 *
 * Drives `place-core` through the generated Kotlin binding to prove the FFI
 * surface lifts/lowers correctly on the JVM before wiring it into the Android
 * app. Uses a representative subset of the golden cases (constructed via the
 * generated data classes, so no JSON dependency) — full fixture parity is
 * proven by the Python client, which replays every case.
 *
 * Run from the crate root (see ../README.md):
 *   kotlinc <binding>.kt GoldenClient.kt -cp jna.jar -include-runtime -d gc.jar
 *   java -cp gc.jar:jna.jar -Djava.library.path=target/debug GoldenClientKt
 */

import uniffi.place_core.Candidate
import uniffi.place_core.EngineException
import uniffi.place_core.Kommune
import uniffi.place_core.PlaceEngine
import uniffi.place_core.PlaceEngineInterface
import uniffi.place_core.ProtectedArea
import uniffi.place_core.Qualifier
import uniffi.place_core.ReverseInput
import uniffi.place_core.SearchCandidate
import java.io.File

private fun check(cond: Boolean, msg: String, fails: MutableList<String>) {
    if (!cond) fails.add(msg)
}

/** Representative golden assertions, run against any engine. */
private fun runChecks(engine: PlaceEngineInterface, label: String, fails: MutableList<String>) {
    check(engine.rulesetVersion() == "1", "[$label] unexpected ruleset version", fails)

    // Reverse: a peak within 100 m is tagged On.
    engine.reverseGeocode(
        ReverseInput(
            toponyms = listOf(Candidate("Galdhøpiggen", "Fjelltopp", 33.0, "aktiv", null)),
            protectedArea = null,
            address = null,
            kommune = null,
            elevationM = null,
        ),
    ).let { d ->
        check(d?.title == "Galdhøpiggen" && d?.qualifier == Qualifier.ON, "[$label] peak On: $d", fails)
    }

    // Reverse: a containing park wins when no toponym; kommune enriches.
    engine.reverseGeocode(
        ReverseInput(
            toponyms = emptyList(),
            protectedArea = ProtectedArea("Saltfjellet–Svartisen nasjonalpark", "Nasjonalpark"),
            address = null,
            kommune = Kommune("Bodø", "Nordland"),
            elevationM = null,
        ),
    ).let { d ->
        check(
            d?.title == "Saltfjellet–Svartisen nasjonalpark" &&
                d?.qualifier == Qualifier.IN_AREA &&
                d?.kommune == "Bodø",
            "[$label] park wins: $d",
            fails,
        )
    }

    // Reverse: kommune fallback carries no qualifier.
    engine.reverseGeocode(
        ReverseInput(emptyList(), null, null, Kommune("Lom", "Innlandet"), null),
    ).let { d ->
        check(
            d?.title == "Lom" && d?.qualifier == null && d?.secondary == "Innlandet",
            "[$label] kommune fallback: $d",
            fails,
        )
    }

    // Forward: exact beats prefix; icon mapping resolves.
    engine.forwardSearch(
        "stor",
        listOf(
            SearchCandidate("Storsteinnes", "Tettsted", null, null),
            SearchCandidate("Stor", "Fjell", null, null),
        ),
    ).let { hits ->
        check(hits.map { it.title } == listOf("Stor", "Storsteinnes"), "[$label] order: ${hits.map { it.title }}", fails)
        check(hits.first().icon == "mountain", "[$label] icon: ${hits.first().icon}", fails)
    }
}

fun main(args: Array<String>) {
    val fails = mutableListOf<String>()

    // 1. Embedded ruleset.
    runChecks(PlaceEngine.withDefaultRuleset(), "default", fails)

    // 3a. Round-trip: an engine built from the ruleset JSON behaves identically.
    val rulesetPath = args.getOrElse(0) { "ruleset.v1.json" }
    val rulesetJson = File(rulesetPath).readText()
    runChecks(PlaceEngine.fromRulesetJson(rulesetJson), "from-json", fails)

    // 3b. Invalid ruleset raises the typed FFI error rather than crashing.
    try {
        PlaceEngine.fromRulesetJson("{ not valid json")
        fails.add("[error] fromRulesetJson accepted invalid JSON")
    } catch (_: EngineException.InvalidRuleset) {
        // expected
    }

    if (fails.isNotEmpty()) {
        System.err.println("FAIL — ${fails.size} case(s) diverged across the FFI boundary:")
        fails.forEach { System.err.println("  $it") }
        kotlin.system.exitProcess(1)
    }
    println("OK — golden + round-trip + error path pass through the UniFFI Kotlin binding")
}

File: packages/place-core/clients/python/golden_client.py
#!/usr/bin/env python3
"""Synthetic UniFFI test client (Python).

Drives `place-core` through the *generated UniFFI binding* — not the Rust API —
and replays the same `golden.json` / `golden_search.json` fixtures the Rust
tests use. If this passes, the FFI surface (records, enums, the PlaceEngine
object) lifts/lowers correctly, so we can trust the bindings before wiring them
into the apps.

Run via ../../clients/run_python_client.sh (it copies the cdylib next to the
generated module first).
"""

import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
CRATE = os.path.normpath(os.path.join(HERE, "..", ".."))
sys.path.insert(0, HERE)

import place_core as pc  # noqa: E402  (must follow sys.path tweak + cdylib copy)

_QUALIFIER = {
    "on": pc.Qualifier.ON,
    "closeTo": pc.Qualifier.CLOSE_TO,
    "atPlace": pc.Qualifier.AT_PLACE,
    "inArea": pc.Qualifier.IN_AREA,
    "near": pc.Qualifier.NEAR,
}


def _candidate(d):
    return pc.Candidate(
        name=d["name"],
        kind=d.get("kind", ""),
        distance_m=float(d["distance_m"]),
        status=d.get("status"),
        secondary=d.get("secondary"),
    )


def _reverse_input(d):
    pa = d.get("protected_area")
    addr = d.get("address")
    komm = d.get("kommune")
    return pc.ReverseInput(
        toponyms=[_candidate(c) for c in d.get("toponyms", [])],
        protected_area=pc.ProtectedArea(name=pa["name"], kind=pa.get("kind")) if pa else None,
        address=pc.Address(text=addr["text"], secondary=addr.get("secondary")) if addr else None,
        kommune=pc.Kommune(name=komm["name"], fylke=komm.get("fylke")) if komm else None,
        elevation_m=d.get("elevation_m"),
    )


def _expected_description(d):
    if d is None:
        return None
    q = d.get("qualifier")
    return pc.LocationDescription(
        title=d["title"],
        qualifier=_QUALIFIER[q] if q else None,
        secondary=d.get("secondary"),
        kommune=d.get("kommune"),
        fylke=d.get("fylke"),
        distance_m=d.get("distance_m"),
        elevation_m=d.get("elevation_m"),
    )


def _search_candidate(d):
    return pc.SearchCandidate(
        name=d["name"],
        kind=d.get("kind", ""),
        distance_m=d.get("distance_m"),
        description=d.get("description"),
    )


def _expected_hit(d):
    return pc.SearchHit(
        index=d["index"],
        title=d["title"],
        description=d.get("description"),
        icon=d["icon"],
    )


def _load(name):
    with open(os.path.join(CRATE, name), encoding="utf-8") as f:
        return json.load(f)


def main():
    engine = pc.PlaceEngine.with_default_ruleset()
    assert engine.ruleset_version() == "1", "binding reached an unexpected ruleset"

    failures = []

    for case in _load("golden.json"):
        got = engine.reverse_geocode(_reverse_input(case["input"]))
        want = _expected_description(case["expect"])
        if got != want:
            failures.append(f"[reverse] {case['name']!r}\n    want: {want}\n    got:  {got}")

    for case in _load("golden_search.json"):
        got = engine.forward_search(case["query"], [_search_candidate(c) for c in case["candidates"]])
        want = [_expected_hit(h) for h in case["expect"]]
        if got != want:
            failures.append(f"[search]  {case['name']!r}\n    want: {want}\n    got:  {got}")

    if failures:
        print(f"FAIL — {len(failures)} case(s) diverged across the FFI boundary:\n")
        print("\n\n".join(failures))
        return 1

    print("OK — golden fixtures pass through the UniFFI Python binding")
    return 0


if __name__ == "__main__":
    sys.exit(main())
